### PR TITLE
[Reviewer: Rob] Add more stats tables

### PR DESCRIPTION
**Pull request by eleanor-merry**
Rob, can you review this change to add a new type of stats table indexed by time and node type. 

Tested live

### CODE REVIEW

File: include/snmp_internal/snmp_counts_by_node_type_table.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+#ifndef SNMP_COUNTS_BY_NODE_TYPE_TABLE_H
+#define SNMP_COUNTS_BY_NODE_TYPE_TABLE_H
+
+#include "snmp_internal/snmp_time_period_and_node_type_table.h"
+
+namespace SNMP
+{
+
+template <class T, int i> class CountsByNodeTypeTableImpl: public ManagedTable<T, std::pair<int, int>>
```
Comment (rkd-msw) [inline include/snmp_internal/snmp_counts_by_node_type_table.h]: `T` and `i` could have more descriptive names - e.g. RowType and column_num?

File: include/snmp_internal/snmp_counts_by_node_type_table.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  typename T::CurrentAndPrevious five_second_i;
+  typename T::CurrentAndPrevious five_minute_i;
+  typename T::CurrentAndPrevious five_second_s;
+  typename T::CurrentAndPrevious five_minute_s;
+  typename T::CurrentAndPrevious five_second_b;
+  typename T::CurrentAndPrevious five_minute_b;
+};
+
+template <class T, int i> class CountsByNodeTypeTable
```
Comment (rkd-msw) [inline include/snmp_internal/snmp_counts_by_node_type_table.h]: Is this used anywhere?

File: include/snmp_internal/snmp_counts_by_node_type_table.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopePrevious5SecondPeriod, NodeTypes::ICSCF) );
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopeCurrent5MinutePeriod, NodeTypes::ICSCF) );
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopePrevious5MinutePeriod, NodeTypes::ICSCF) );
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopePrevious5SecondPeriod, NodeTypes::BGCF) );
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopeCurrent5MinutePeriod, NodeTypes::BGCF) );
+    this->add( std::pair<int, int>(TimePeriodIndexes::scopePrevious5MinutePeriod, NodeTypes::BGCF) );
+  }
+
+protected:
+  T* new_row(struct std::pair<int, int> indexes)
```
Comment (rkd-msw) [inline include/snmp_internal/snmp_counts_by_node_type_table.h]: This feels like a lot of boilerplate, which is a shame. I've done a bit of tweaking and think I've got a way to reduce it - I'll send a new PR against this branch.

File: src/snmp_ip_count_table.cpp [hunk 4]
```diff
@@ -130,7 +130,7 @@ class IPCountTableImpl: public ManagedTable<IPCountRow, std::string>, public IPC
     }
     else if (inet_pton(AF_INET6, ip.c_str(), &v6) == 1)
     {
-      return new IPCountRow(v4);
+      return new IPCountRow(v6);
```
Comment (rkd-msw) [inline src/snmp_ip_count_table.cpp]: In general, we should probably put this kind of change in a separate PR - makes it easier to track for back-application.

File: include/snmp_internal/snmp_time_period_and_node_type_table.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+template <class T> class TimeAndNodeTypeBasedRow : public TimeBasedRow<T>
+{
+public:
+  // Constructor, takes ownership of the View*.
+  TimeAndNodeTypeBasedRow(int time_index, int type_index, typename TimeBasedRow<T>::View* view) :
+    TimeBasedRow<T>(time_index, view),
+    _type_index(type_index)
+  {
+    // Add index for the node type
```
Comment (rkd-msw) [inline include/snmp_internal/snmp_time_period_and_node_type_table.h]: Worth a comment to note that the time index is added in the base constructor?

File: include/snmp_internal/snmp_time_period_and_node_type_table.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+#define SNMP_TIME_PERIOD_AND_NODE_TYPE_TABLE_H
+
+// This file contains the base infrastructure for SNMP tables 
+// which are indexed by time period and node type.
+namespace SNMP
+{
+
+enum NodeTypes
+{
+  SCSCF = 0,
```
Comment (rkd-msw) [inline include/snmp_internal/snmp_time_period_and_node_type_table.h]: Worth adding the full set now, so that the gaps don't look odd?

**Comment by rkd-msw** [comment]
Looks good - I know you're relatively swamped today, so I'll have a crack at addressing my biggest comment (boilerplate) myself.

**Comment by rkd-msw** [comment]
Boilerplate PR is #331.

File: include/snmp_internal/snmp_counts_by_node_type_table.h (hunk 1 above)
Comment (bossmc) [inline include/snmp_internal/snmp_counts_by_node_type_table.h]: Can't we learn `i` from `T`?  Presumably as a const class function on `T`?  Otherwise there's nothing (asides from run-time crashes) to prevent us from diverging these two sizes.

Comment (rkd-msw) [inline include/snmp_internal/snmp_counts_by_node_type_table.h]: It'd be nice to do something more fundamental - having `SNMP::Row` subclasses classes expose their index types and a count of non-index columns, which the `SNMP::Table` could then learn (in lieu of its last three constructor arguments).
